### ISSUE TITLE
reshape momentum inside the integrator?

### ISSUE THREAD

**Issue by rlouf** [rlouf/mcx#7]
There is a tension inside the HMC algorithm between the momentum generator, which generates 1D arrays, and the log-probability functions which takes tuples of array. Since the integrator step involves computations that mix for, we either need to work with flattened variables or reshape the momentum and use `jax.flatten_util.ravel_pytree` to obtain an unraveling function. Here are the tradeoffs:

## Flattening variables

This can be easily achieved by composing the unraveling function with the logpdf. We can then use variables as a concatenated 1D Array.

This leads to the following extra operations:

- ˋnum_stages` unraveling per integration steps: N_iter * num_stages * num_steps

**Pros**

- The inference code only deals with 1D Arrays. In a way it is cleaner.

**Cons**

- It may make extensions such as discontinuous HMC and Metropolis-within-Gibbs more complex down the line as we would have to carry the shape of each variable around.
- It adds extra steps for users who lay want to use the inference engine with custom logpdfs; it makes the user interface complicated.
- We need to pass an unraveling function to the integrator steps; it is harder to justify as it is far from the source of the issue.

## Unraveling momentum during inference

This can be done at several stages. Either in the momentum generator (like Numpyro) or the integrator. It seems kind of awkward to do it in the momentum generator: it becomes disconnected from the place it is actually needed and will look confusing to someone first encountering the code.

**Pro**

- Transformation is done at a place where we see the actual data. No need to add and extra step to guess the shape at initialization.
- Advanced users only have to worry about providing an unraveling function, which is easier to do and to understand.

**Cons**

- One op to generate the momentum, one op to compute the kinetic energy
- `num_stages` ravelling / unravelling operations per integration step (cf JAX' [tree_multimap implementation](https://github.com/google/jax/blob/1006758b943191133e1ffb4dd5003deb594b2f9a/jax/tree_util.py#L123)
- We need to replace all operations in the integrators with `tree_multimap` ops. Slightly more verbose.

**Comment by rlouf** [comment]
I settled on the second option for now. The momentum is an array because of the existence of the mas matrix; this may be a quirk of the euclidean metric: RHMC's momentum generator and kinetic energy is a function of the position, for instance. We should thus not carry this specificity around in the code.

Note that this is a potential source of performance improvement, although I suspect the raveling/unraveling takes substantially less time than logpdf and gradient of the logpdf evaluation.
